### ISSUE TITLE
Are you willing to consider dropping python2.7 support ?

### ISSUE THREAD

**Issue by ticosax** [biosustain/potion#163]

**Comment by lyschoening** [comment]
Do you have a change in mind that would necessitate dropping Python 2.7 support or that would be more difficult to implement in Python 2.7? At this point, I find that there is not much overhead in maintaining the Python 2 version of the code for this project, so I would not drop support before EOY or before it ceases to be a requirement for official Flask extensions:

> An extension currently has to support Python 3.4 and newer and 2.7.
http://flask.pocoo.org/docs/1.0/extensiondev/#approved-extensions

**Comment by lyschoening** [comment]
I'm simply cautious about dropping support without good reason. In my view, if there is code that is currently harder to maintain because of support for Python 2 and there is a need to touch that code, that is a good reason to drop support a few months earlier. It sounds as if you are looking to refactor the library significantly and that could be a very good reason.

**Comment by ticosax** [comment]
We are using potion for our production workload. If you welcome contributions we will do more of them, to make sure potion stays relevant for us, as we need to grow.
Also if we are able to make option attractive again, we might be able to welcome more developers willing to help.
With a bit of anticipation and communication towards a release plan, we should be able to break backward compatibility with a new major version. Explaining for users stuck with python2.7 that potion maintenance will be discontinued for them.
I think it is time to consider dropping python2.7, if you agree, we could discuss here how we get there, in terms of release schedule ?

**Comment by lyschoening** [comment]
I am open to this idea. Give me a few days to discuss this internally, as
we are currently reviewing how to structure the maintenance of some of our
open source projects. Perhaps we can hop on a chat next week to address
this and your other open issues.

**Comment by lyschoening** [comment]
@ticosax Thanks for your patience. Yes, dropping Python 2.7 support with an upcoming release is on the table. #161 and #162 are also interesting – and overdue – initiatives. What would be a good way to discuss the development that you are looking to have done on Potion to keep it relevant for your products?

**Comment by ticosax** [comment]
Thanks @lyschoening 

> What would be a good way to discuss the development that you are looking to have done on Potion to keep it relevant for your products?

I suggest to discuss our intentions on separate github issues. Once we agree, then we'll know we can make the effort to bring improvements knowing that it won't be forgotten.
I expect from the maintainer of the project like you to use your general knowledge and experience to challenge the proposals and keep the project on good track, with a consistent and developer-friendly API, and with the level of quality one would expect to run critical products. In short you'll be a decision maker, which is a tough role !

Be prepared for some controversial debates  like dropping some ORM support and encouraging maintenance of them as third party library :smile:
That's the kind of things I'd like to discuss, in order to keep the maintenance as sustainable as possible.

Also, to give you some context, coming from django world, most of my technical  and governance inspirations will come from django-rest-framework, which I believe choose the right abstractions in order to expose models on an Rest API. This project grow a lot in the last years with an active community behind supporting it, including a rich eco-system of third party libraries. potion could become the equivalent for flask.

**Comment by ticosax** [comment]
@lyschoening just for your Information, we are maintaining our own fork under `infarm/potion`. We aim to submit every patches to the upstream repository and keep the delta as small as possible.
Have you considered inviting external maintainers or transferring the project to another organization ? https://github.com/jazzband could be a good option for instance.

**Comment by lyschoening** [comment]
@ticosax Would you be interested in taking over the project at Infarm?

**Comment by ticosax** [comment]
Yes, but to be honest I a firm believer that it should become a community project, it means we will probably transfer ownership to jazzband or similar.
Out of curiosity, are you yourself still using potion ?

**Comment by lyschoening** [comment]
> Out of curiosity, are you yourself still using potion ?

No. I can no longer speak for DTU Biosustain, as I have since moved on to another company and am only a part of the organization as a maintainer of potion. Due to changing needs, we did drop potion in favor graphene (GraphQL) in my time at DTU Biosustain and I don't know that this has changed. 

Jazzband may be the right option. I will have to reach out to Jazzband as I am not certain I will be able to transfer the biosustain/potion repository. I'll either have to fork the repository or create a new one. As you have correctly observed, maintenance under the current repository is certainly not the way forward.

**Comment by ticosax** [comment]
I see, that explain a lot.
I suggest, I take care of transferring the project to jazzband, as it can be a tedious process.
For that you would only need to transfer ownership to infarm first.
If you agree, I'm going to delete our fork, to make some clearance for your transfer.
Please confirm.

**Comment by lyschoening** [comment]
@ticosax Due to internal politics I can't transfer the repository away from the DTU Biosustain organization. Otherwise, I would likely have done so much earlier. We can make a copy of the repository and update this repository with a message pointing to the new location. This of course means that we'll lose the stars and any current issues. 

I will do another push to allow a direct transfer to another organization, but I cannot guarantee its success.

**Comment by ticosax** [comment]
I see, so it means we won't have access to the namespace on pypi either, I assume.
You should advertise the state of the project on the README anyway, to help contributors managing their expectation.
We don't care much about maintaining this project on the long run. We are also transitioning to GraphQL, In the mean time we want to make sure potion stays at quality level it should be, Like supporting python3.7 and eventually bring some performance optimization here and there. Low hanging fruit only.

**Comment by lyschoening** [comment]
> I see, so it means we won't have access to the namespace on pypi either, I assume.

I do have access to the namespace on PyPI and can share that with other maintainers. Only the access to the GitHub repository was retained, so all that would be lost with a transfer is the stars and automatic redirect. It would be good to move the project over to Jazzband regardless.

> You should advertise the state of the project on the README anyway, to help contributors managing their expectation.
> We don't care much about maintaining this project on the long run. We are also transitioning to GraphQL, In the mean time we want to make sure potion stays at quality level it should be, Like supporting python3.7 and eventually bring some performance optimization here and there. Low hanging fruit only.

That is fair enough. Other than supporting OpenAPI I don't think there is very much to be done anyway. It's a fine opinionated REST API framework for Flask + single page application pairings that need a simple permission system and can benefit from the TypeScript or Python clients. GraphQL covers the pains this library was written to solve and much more, so anyone considering Potion should first look if they can find comparable tooling for GQL, e.g. Graphene-SQLAlchemy.

I will update the README to indicate that the project essentially in maintenance mode.
